<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Pattern matching and searching · BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../"><img class="logo" src="../assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><a class="toctext" href="../symbols/">Biological Symbols</a></li><li><a class="toctext" href="../types/">BioSequences Types</a></li><li><a class="toctext" href="../construction/">Constructing sequences</a></li><li><a class="toctext" href="../transforms/">Indexing &amp; modifying sequences</a></li><li><a class="toctext" href="../predicates/">Predicates</a></li><li><a class="toctext" href="../random/">Random sequences</a></li><li class="current"><a class="toctext" href>Pattern matching and searching</a><ul class="internal"><li><a class="toctext" href="#Exact-search-1">Exact search</a></li><li><a class="toctext" href="#Approximate-search-1">Approximate search</a></li><li><a class="toctext" href="#Regular-expression-search-1">Regular expression search</a></li><li><a class="toctext" href="#Position-weight-matrix-search-1">Position weight matrix search</a></li></ul></li><li><a class="toctext" href="../demultiplexer/">Sequence Demultipllexing</a></li><li><a class="toctext" href="../composition/">Sequence composition</a></li><li><a class="toctext" href="../iteration/">Iteration</a></li><li><a class="toctext" href="../counting/">Counting</a></li><li><a class="toctext" href="../io/">I/O</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Pattern matching and searching</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/BioSequences.jl/blob/master/docs/src/sequence_search.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Pattern matching and searching</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Sequence-search-1" href="#Sequence-search-1">Sequence search</a></h1><p>Three kinds of on-line search functions are provided:</p><ol><li>Exact search</li><li>Approximate search</li><li>Regular expression search</li></ol><p>These are all specialized for biological sequences and ambiguities of symbols are considered.</p><h2><a class="nav-anchor" id="Exact-search-1" href="#Exact-search-1">Exact search</a></h2><p>Exact search functions search for an occurrence of the query symbol or sequence.</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACAGCGTAGCT&quot;;

julia&gt; findfirst(DNA_G, seq)
4

julia&gt; query = dna&quot;AGC&quot;;

julia&gt; findfirst(query, seq)
3:5

julia&gt; findlast(query, seq)
8:10
</code></pre><p>These search functions take ambiguous symbols into account. That is, if two symbols are compatible (e.g. <code>DNA_A</code> and <code>DNA_N</code>), they match when searching an occurrence. In the following example, &#39;N&#39; is a wild card that matches any symbols.</p><pre><code class="language-julia-repl">julia&gt; findfirst(dna&quot;CGT&quot;, dna&quot;ACNT&quot;)  # &#39;N&#39; matches &#39;G&#39;
2:4

julia&gt; findfirst(dna&quot;CNT&quot;, dna&quot;ACGT&quot;)  # &#39;G&#39; matches &#39;N&#39;
2:4
</code></pre><p>The exception to this behaviour is if you are finding a single &#39;character&#39;, in which case an ambiguous symbol is matched exactly:</p><pre><code class="language-julia-repl">julia&gt; findfirst(DNA_N, dna&quot;ACNT&quot;)
3
</code></pre><p>The exact sequence search needs a preprocessing phase of query sequence before the searching phase. This would be fast enough for most search applications. But when searching a query sequence to many target sequences, caching the result of preprocessing may save time. You can do this by creating an <code>ExactSearchQuery</code> object and re-use it for each search:</p><pre><code class="language-julia-repl">julia&gt; query = ExactSearchQuery(dna&quot;ATT&quot;);

julia&gt; findfirst(query, dna&quot;ATTTATT&quot;)
1:3

julia&gt; findlast(query, dna&quot;ATTTATT&quot;)
5:7
</code></pre><h2><a class="nav-anchor" id="Approximate-search-1" href="#Approximate-search-1">Approximate search</a></h2><p>The approximate search is similar to the exact search but allows a specific number of errors. That is, it tries to find a subsequence of the target sequence within a specific <a href="https://en.wikipedia.org/wiki/Levenshtein_distance">Levenshtein distance</a> of the query sequence:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACAGCGTAGCT&quot;;

julia&gt; approxsearch(seq, dna&quot;AGGG&quot;, 0)  # nothing matches with no errors
0:-1

julia&gt; approxsearch(seq, dna&quot;AGGG&quot;, 1)  # seq[3:6] matches with one error
3:6

julia&gt; approxsearch(seq, dna&quot;AGGG&quot;, 2)  # seq[1:4] matches with two errors
1:4
</code></pre><p>Like the exact search functions, four kinds of functions (<code>approxsearch</code>, <code>approxsearchindex</code>, <code>approxrsearch</code>, and <code>approxrsearchindex</code>) are available:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACAGCGTAGCT&quot;; pat = dna&quot;AGGG&quot;;

julia&gt; approxsearch(seq, pat, 2)        # return the range (forward)
1:4

julia&gt; approxsearchindex(seq, pat, 2)   # return the starting index (forward)
1

julia&gt; approxrsearch(seq, pat, 2)       # return the range (backward)
8:11

julia&gt; approxrsearchindex(seq, pat, 2)  # return the starting index (backward)
8
</code></pre><p>Preprocessing can be cached in an <code>ApproximateSearchQuery</code> object:</p><pre><code class="language-julia-repl">julia&gt; query = ApproximateSearchQuery(dna&quot;AGGG&quot;);

julia&gt; approxsearch(dna&quot;AAGAGG&quot;, query, 1)
2:5

julia&gt; approxsearch(dna&quot;ACTACGT&quot;, query, 2)
4:6
</code></pre><h2><a class="nav-anchor" id="Regular-expression-search-1" href="#Regular-expression-search-1">Regular expression search</a></h2><p>Query patterns can be described in regular expressions. The syntax supports a subset of Perl and PROSITE&#39;s notation.</p><p>The Perl-like syntax starts with <code>biore</code> (BIOlogical REgular expression) and ends with a symbol option: &quot;dna&quot;, &quot;rna&quot; or &quot;aa&quot;. For example, <code>biore&quot;A+&quot;dna</code> is a regular expression for DNA sequences and <code>biore&quot;A+&quot;aa</code> is for amino acid sequences. The symbol options can be abbreviated to its first character: &quot;d&quot;, &quot;r&quot; or &quot;a&quot;, respectively.</p><p>Here are examples of using the regular expression for <code>BioSequence</code>s:</p><pre><code class="language-julia-repl">julia&gt; match(biore&quot;A+C*&quot;dna, dna&quot;AAAACC&quot;)
RegexMatch(&quot;AAAACC&quot;)

julia&gt; match(biore&quot;A+C*&quot;d, dna&quot;AAAACC&quot;)
RegexMatch(&quot;AAAACC&quot;)

julia&gt; occursin(biore&quot;A+C*&quot;dna, dna&quot;AAC&quot;)
true

julia&gt; occursin(biore&quot;A+C*&quot;dna, dna&quot;C&quot;)
false
</code></pre><p><code>match</code> will return a <code>RegexMatch</code> if a match is found, otherwise it will return <code>nothing</code> if no match is found.</p><p>The table below summarizes available syntax elements.</p><table><tr><th style="text-align: center">Syntax</th><th style="text-align: left">Description</th><th style="text-align: left">Example</th></tr><tr><td style="text-align: center"><code>|</code></td><td style="text-align: left">alternation</td><td style="text-align: left"><code>&quot;A|T&quot;</code> matches <code>&quot;A&quot;</code> and <code>&quot;T&quot;</code></td></tr><tr><td style="text-align: center"><code>*</code></td><td style="text-align: left">zero or more times repeat</td><td style="text-align: left"><code>&quot;TA*&quot;</code> matches <code>&quot;T&quot;</code>, <code>&quot;TA&quot;</code> and <code>&quot;TAA&quot;</code></td></tr><tr><td style="text-align: center"><code>+</code></td><td style="text-align: left">one or more times repeat</td><td style="text-align: left"><code>&quot;TA+&quot;</code> matches <code>&quot;TA&quot;</code> and <code>&quot;TAA&quot;</code></td></tr><tr><td style="text-align: center"><code>?</code></td><td style="text-align: left">zero or one time</td><td style="text-align: left"><code>&quot;TA?&quot;</code> matches <code>&quot;T&quot;</code> and <code>&quot;TA&quot;</code></td></tr><tr><td style="text-align: center"><code>{n,}</code></td><td style="text-align: left"><code>n</code> or more times repeat</td><td style="text-align: left"><code>&quot;A{3,}&quot;</code> matches <code>&quot;AAA&quot;</code> and <code>&quot;AAAA&quot;</code></td></tr><tr><td style="text-align: center"><code>{n,m}</code></td><td style="text-align: left"><code>n</code>-<code>m</code> times repeat</td><td style="text-align: left"><code>&quot;A{3,5}&quot;</code> matches <code>&quot;AAA&quot;</code>, <code>&quot;AAAA&quot;</code> and <code>&quot;AAAAA&quot;</code></td></tr><tr><td style="text-align: center"><code>^</code></td><td style="text-align: left">the start of the sequence</td><td style="text-align: left"><code>&quot;^TAN*&quot;</code> matches <code>&quot;TATGT&quot;</code></td></tr><tr><td style="text-align: center"><code>$</code></td><td style="text-align: left">the end of the sequence</td><td style="text-align: left"><code>&quot;N*TA$&quot;</code> matches <code>&quot;GCTA&quot;</code></td></tr><tr><td style="text-align: center"><code>(...)</code></td><td style="text-align: left">pattern grouping</td><td style="text-align: left"><code>&quot;(TA)+&quot;</code> matches <code>&quot;TA&quot;</code> and <code>&quot;TATA&quot;</code></td></tr><tr><td style="text-align: center"><code>[...]</code></td><td style="text-align: left">one of symbols</td><td style="text-align: left"><code>&quot;[ACG]+&quot;</code> matches <code>&quot;AGGC&quot;</code></td></tr></table><p><code>eachmatch</code> and <code>findfirst</code> are also defined like usual strings:</p><pre><code class="language-julia-repl">julia&gt; collect(matched(x) for x in eachmatch(biore&quot;TATA*?&quot;d, dna&quot;TATTATAATTA&quot;)) # overlap
4-element Array{LongSequence{DNAAlphabet{4}},1}:
 TAT  
 TAT
 TATA
 TATAA

julia&gt; collect(matched(x) for x in eachmatch(biore&quot;TATA*&quot;d, dna&quot;TATTATAATTA&quot;, false)) # no overlap
2-element Array{LongSequence{DNAAlphabet{4}},1}:
 TAT  
 TATAA

julia&gt; findfirst(biore&quot;TATA*&quot;d, dna&quot;TATTATAATTA&quot;)
1:3

julia&gt; findfirst(biore&quot;TATA*&quot;d, dna&quot;TATTATAATTA&quot;, 2)
4:8
</code></pre><p>Noteworthy differences from strings are:</p><ul><li>Ambiguous characters match any compatible characters (e.g. <code>biore&quot;N&quot;d</code> is equivalent to <code>biore&quot;[ACGT]&quot;d</code>).</li><li>Whitespaces are ignored (e.g. <code>biore&quot;A C G&quot;d</code> is equivalent to <code>biore&quot;ACG&quot;d</code>).</li></ul><p>The PROSITE notation is described in <a href="https://prosite.expasy.org/scanprosite/scanprosite_doc.html">ScanProsite - user manual</a>. The syntax supports almost all notations including the extended syntax. The PROSITE notation starts with <code>prosite</code> prefix and no symbol option is needed because it always describes patterns of amino acid sequences:</p><pre><code class="language-julia-repl">julia&gt; match(prosite&quot;[AC]-x-V-x(4)-{ED}&quot;, aa&quot;CPVPQARG&quot;)
RegexMatch(&quot;CPVPQARG&quot;)

julia&gt; match(prosite&quot;[AC]xVx(4){ED}&quot;, aa&quot;CPVPQARG&quot;)
RegexMatch(&quot;CPVPQARG&quot;)
</code></pre><h2><a class="nav-anchor" id="Position-weight-matrix-search-1" href="#Position-weight-matrix-search-1">Position weight matrix search</a></h2><p>A motif can also be specified using <a href="https://en.wikipedia.org/wiki/Position_weight_matrix">position weight matrix</a> (PWM) in a probabilistic way. <code>search(seq, pwm, threshold)</code> method searches for the first position in the sequence where a score calculated using the PWM is greater than or equal to the threshold. More formally, denoting the sequence as <span>$S$</span> and the PWM value of symbol <span>$s$</span> at position <span>$j$</span> as <span>$M_{s,j}$</span>, the score starting from a position <span>$p$</span> is defined as</p><div>\[\operatorname{score}(S, p) = \sum_{i=1}^L M_{S[p+i-1],i}\]</div><p>and <code>search(S, M, t)</code> returns the smallest <span>$p$</span> that satisfies <span>$\operatorname{score}(S, p) \ge t$</span>.</p><p>There are two kinds of matrices in this package: <code>PFM</code> and <code>PWM</code>. The <code>PFM</code> type is a position frequency matrix and stores symbol frequencies for each position. The <code>PWM</code> is a position weight matrix and stores symbol scores for each position. You can create a <code>PFM</code> from a set of sequences with the same length and then create a <code>PWM</code> from the <code>PFM</code> object.</p><pre><code class="language-julia-repl">julia&gt; kmers = DNAMer.([&quot;TTA&quot;, &quot;CTA&quot;, &quot;ACA&quot;, &quot;TCA&quot;, &quot;GTA&quot;])
5-element Array{Mer{DNAAlphabet{2},3},1}:
 TTA
 CTA
 ACA
 TCA
 GTA

julia&gt; pfm = PFM(kmers)  # sequence set =&gt; PFM
4×3 PFM{DNA,Int64}:
 A  1  0  5
 C  1  2  0
 G  1  0  0
 T  2  3  0

julia&gt; pwm = PWM(pfm)  # PFM =&gt; PWM
4×3 PWM{DNA,Float64}:
 A -0.321928 -Inf       2.0
 C -0.321928  0.678072 -Inf
 G -0.321928 -Inf      -Inf
 T  0.678072  1.26303  -Inf

julia&gt; pwm = PWM(pfm .+ 0.01)  # add pseudo counts to avoid infinite values
4×3 PWM{DNA,Float64}:
 A -0.319068 -6.97728   1.99139
 C -0.319068  0.673772 -6.97728
 G -0.319068 -6.97728  -6.97728
 T  0.673772  1.25634  -6.97728

julia&gt; pwm = PWM(pfm .+ 0.01, prior=[0.2, 0.3, 0.3, 0.2])  # GC-rich prior
4×3 PWM{DNA,Float64}:
 A  0.00285965 -6.65535   2.31331
 C -0.582103    0.410737 -7.24031
 G -0.582103   -7.24031  -7.24031
 T  0.9957      1.57827  -6.65535
</code></pre><p>The <span>$PWM_{s,j}$</span> matrix is computed from <span>$PFM_{s,j}$</span> and the prior probability <span>$p(s)$</span> as follows ([Wasserman2004]):</p><div>\[\begin{align}
    PWM_{s,j} &amp;= \log_2 \frac{p(s,j)}{p(s)} \\
    p(s,j)  &amp;= \frac{PFM_{s,j}}{\sum_{s&#39;} PFM_{s&#39;,j}}.
\end{align}\]</div><p>[Wasserman2004]: https://doi.org/10.1038/nrg1315</p><footer><hr/><a class="previous" href="../random/"><span class="direction">Previous</span><span class="title">Random sequences</span></a><a class="next" href="../demultiplexer/"><span class="direction">Next</span><span class="title">Sequence Demultipllexing</span></a></footer></article></body></html>
