<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>BioSequences Types · BioSequences.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="BioSequences.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">BioSequences.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../symbols/">Biological Symbols</a></li><li class="is-active"><a class="tocitem" href>BioSequences Types</a><ul class="internal"><li><a class="tocitem" href="#The-abstract-BioSequence-1"><span>The abstract BioSequence</span></a></li><li><a class="tocitem" href="#Long-Sequences-1"><span>Long Sequences</span></a></li><li><a class="tocitem" href="#Kmers-and-Skipmers-1"><span>Kmers &amp; Skipmers</span></a></li><li><a class="tocitem" href="#Reference-sequences-1"><span>Reference sequences</span></a></li><li class="toplevel"><a class="tocitem" href="#Alphabet-types-1"><span>Alphabet types</span></a></li></ul></li><li><a class="tocitem" href="../construction/">Constructing sequences</a></li><li><a class="tocitem" href="../transforms/">Indexing &amp; modifying sequences</a></li><li><a class="tocitem" href="../predicates/">Predicates</a></li><li><a class="tocitem" href="../random/">Random sequences</a></li><li><a class="tocitem" href="../sequence_search/">Pattern matching and searching</a></li><li><a class="tocitem" href="../demultiplexer/">Sequence Demultipllexing</a></li><li><a class="tocitem" href="../composition/">Sequence composition</a></li><li><a class="tocitem" href="../iteration/">Iteration</a></li><li><a class="tocitem" href="../counting/">Counting</a></li><li><a class="tocitem" href="../io/">I/O</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>BioSequences Types</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>BioSequences Types</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/BioSequences.jl/blob/master/docs/src/types.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Sequence-Types-1"><a class="docs-heading-anchor" href="#Sequence-Types-1">Sequence Types</a><a class="docs-heading-anchor-permalink" href="#Sequence-Types-1" title="Permalink"></a></h1><p>BioSequences exports an abstract <code>BioSequence</code> type, and several concrete sequence types which inherit from it.</p><h2 id="The-abstract-BioSequence-1"><a class="docs-heading-anchor" href="#The-abstract-BioSequence-1">The abstract BioSequence</a><a class="docs-heading-anchor-permalink" href="#The-abstract-BioSequence-1" title="Permalink"></a></h2><p>BioSequences provides an abstract type called a <code>BioSequence{A&lt;:Alphabet}</code>. This abstract type, and the methods and traits is supports, allows for many algorithms in BioSequences to be written as generically as possible, thus reducing the amount of code to read and understand, whilst maintaining high performance when such code is compiled for a concrete BioSequence subtype. Additionally, it allows new types to be implemented that are fully compatible with the rest of BioSequences, providing that key methods or traits are defined).</p><p>This abstract type is parametric over concrete types of <code>Alphabet</code>, which define the range of symbols permitted in the sequence.</p><p>Some aliases are also provided for your convenience:</p><table><tr><th style="text-align: left">Type alias</th><th style="text-align: left">Type</th></tr><tr><td style="text-align: left"><code>NucleotideSeq</code></td><td style="text-align: left"><code>BioSequence{&lt;:NucleicAcidAlphabet}</code></td></tr><tr><td style="text-align: left"><code>AminoAcidSeq</code></td><td style="text-align: left"><code>BioSequence{AminoAcidAlphabet}</code></td></tr></table><p>Any concrete sequence type compatible with BioSequences must inherit from <code>BioSequence{A}</code>, where <code>A</code> is the alphabet of the concrete sequence type. It must also have the following methods defined for it:</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.encoded_data" href="#BioSequences.encoded_data"><code>BioSequences.encoded_data</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Return the data member of <code>seq</code> that stores the encoded sequence data.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/biosequence/biosequence.jl#L32-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Base.length-Tuple{BioSequence}" href="#Base.length-Tuple{BioSequence}"><code>Base.length</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Get the length of a biological sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/biosequence/biosequence.jl#L21">source</a></section></article><p>If these requirements are satisfied, the following key traits and methods backing the BioSequences interface, should be defined already for the sequence type.</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.encoded_data_type" href="#BioSequences.encoded_data_type"><code>BioSequences.encoded_data_type</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the vector of bits storing a sequences packed encoded elements.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/biosequence/biosequence.jl#L52">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.encoded_data_eltype" href="#BioSequences.encoded_data_eltype"><code>BioSequences.encoded_data_eltype</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the element type of the vector of bits storing a sequences packed encoded elements.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/biosequence/biosequence.jl#L55">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.Alphabet-Tuple{BioSequence}" href="#BioSequences.Alphabet-Tuple{BioSequence}"><code>BioSequences.Alphabet</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Return the <code>Alpahbet</code> type that defines the biological symbols allowed for <code>seq</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/biosequence/biosequence.jl#L66">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSymbols.alphabet-Tuple{BioSequence}" href="#BioSymbols.alphabet-Tuple{BioSequence}"><code>BioSymbols.alphabet</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Gets the alphabet encoding of a given BioSequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/longsequences/longsequence.jl#L68">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.BitsPerSymbol" href="#BioSequences.BitsPerSymbol"><code>BioSequences.BitsPerSymbol</code></a> — <span class="docstring-category">Type</span></header><section><div><p>The number of bits required to represent a packed symbol in a vector of bits.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/alphabet.jl#L40-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.bits_per_symbol" href="#BioSequences.bits_per_symbol"><code>BioSequences.bits_per_symbol</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Get the number of bits each symbol packed into a BioSequence uses, as an integer value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/biosequence/biosequence.jl#L75">source</a></section></article><p>As a result, the vast majority of methods described in the rest of this manual should work out of the box for the concrete sequence type. But they can always be overloaded if needed.</p><h2 id="Long-Sequences-1"><a class="docs-heading-anchor" href="#Long-Sequences-1">Long Sequences</a><a class="docs-heading-anchor-permalink" href="#Long-Sequences-1" title="Permalink"></a></h2><p>Many genomics scripts and tools benefit from an efficient general purpose sequence type that allows you to create and edit sequences. In BioSequences, the <code>LongSequence</code> type fills this requirement.</p><p><code>LongSequence{A&lt;:Alphabet} &lt;: BioSequence{A}</code> is parameterized by a concrete <code>Alphabet</code> type <code>A</code> that defines the domain (or set) of biological symbols permitted. For example, <code>AminoAcidAlphabet</code> is associated with <code>AminoAcid</code> and hence an object of the <code>LongSequence{AminoAcidAlphabet}</code> type represents a sequence of amino acids.  Symbols from multiple alphabets can&#39;t be intermixed in one sequence type.</p><p>The following table summarizes common LongSequence types that have been given aliases for convenience.</p><table><tr><th style="text-align: left">Type</th><th style="text-align: left">Symbol type</th><th style="text-align: left">Type alias</th></tr><tr><td style="text-align: left"><code>LongSequence{DNAAlphabet{4}}</code></td><td style="text-align: left"><code>DNA</code></td><td style="text-align: left"><code>LongDNASeq</code></td></tr><tr><td style="text-align: left"><code>LongSequence{RNAAlphabet{4}}</code></td><td style="text-align: left"><code>RNA</code></td><td style="text-align: left"><code>LongRNASeq</code></td></tr><tr><td style="text-align: left"><code>LongSequence{AminoAcidAlphabet}</code></td><td style="text-align: left"><code>AminoAcid</code></td><td style="text-align: left"><code>LongAminoAcidSeq</code></td></tr><tr><td style="text-align: left"><code>LongSequence{CharAlphabet}</code></td><td style="text-align: left"><code>Char</code></td><td style="text-align: left"><code>LongCharSeq</code></td></tr></table><p>The <code>LongDNASeq</code> and <code>LongRNASeq</code> aliases use a DNAAlphabet{4}, which means the sequence may store ambiguous nucleotides. If you are sure that nucleotide sequences store unambiguous nucleotides only, you can reduce the memory required by sequences by using a slightly different parameter: <code>DNAAlphabet{2}</code> is an alphabet that uses two bits per base and limits to only unambiguous nucleotide symbols (ACGT in DNA and ACGU in RNA). Replacing <code>LongSequence{DNAAlphabet{4}}</code> in your code with <code>LongSequence{DNAAlphabet{2}}</code> is all you need to do in order to benefit. Some computations that use bitwise operations will also be dramatically faster.</p><h2 id="Kmers-and-Skipmers-1"><a class="docs-heading-anchor" href="#Kmers-and-Skipmers-1">Kmers &amp; Skipmers</a><a class="docs-heading-anchor-permalink" href="#Kmers-and-Skipmers-1" title="Permalink"></a></h2><h3 id="Kmers-1"><a class="docs-heading-anchor" href="#Kmers-1">Kmers</a><a class="docs-heading-anchor-permalink" href="#Kmers-1" title="Permalink"></a></h3><p>Bioinformatic analyses make extensive use of kmers. Kmers are contiguous sub-strings of k nucleotides of some ref sequence. </p><p>They are used extensively in bioinformatic analyses as an informational unit. This concept popularised by short read assemblers.  Analyses within the kmer space benefit from a simple formulation of the sampling problem and direct in-hash comparisons.</p><p>BioSequences provides the following types to represent Kmers, unlike some sequence types, they are immutable.</p><h4 id="Mer{A:NucleicAcidAlphabet{2},K}-1"><a class="docs-heading-anchor" href="#Mer{A:NucleicAcidAlphabet{2},K}-1"><code>Mer{A&lt;:NucleicAcidAlphabet{2},K}</code></a><a class="docs-heading-anchor-permalink" href="#Mer{A:NucleicAcidAlphabet{2},K}-1" title="Permalink"></a></h4><p>Represents a substring of <code>K</code> DNA or RNA nucleotides (depending on the alphabet A). This type represents the sequence using a single <code>UInt64</code>, and so the maximum possible sequence possible - the largest <code>K</code> possible, is <code>32</code>. We recommend using <code>31</code> rather than <code>32</code> however, as odd <code>K</code> values avoid palindromic mers which can be problematic for some algorithms.</p><h4 id="BigMer{A:NucleicAcidAlphabet{2},K}-1"><a class="docs-heading-anchor" href="#BigMer{A:NucleicAcidAlphabet{2},K}-1"><code>BigMer{A&lt;:NucleicAcidAlphabet{2},K}</code></a><a class="docs-heading-anchor-permalink" href="#BigMer{A:NucleicAcidAlphabet{2},K}-1" title="Permalink"></a></h4><p>Represents a substring of <code>K</code> DNA or RNA nucleotides (depending on the alphabet A). This type represents the sequence using a single <code>UInt128</code>, and so the maximum possible sequence possible - the largest <code>K</code> possible, is <code>64</code>. We recommend using <code>63</code> rather than <code>64</code> however, as odd <code>K</code> values avoid palindromic mers which can be problematic for some algorithms.</p><h4 id="AbstractMer{A:NucleicAcidAlphabet{2},K}-1"><a class="docs-heading-anchor" href="#AbstractMer{A:NucleicAcidAlphabet{2},K}-1"><code>AbstractMer{A&lt;:NucleicAcidAlphabet{2},K}</code></a><a class="docs-heading-anchor-permalink" href="#AbstractMer{A:NucleicAcidAlphabet{2},K}-1" title="Permalink"></a></h4><p>This abstract type is just a type that unifies the <code>Mer</code> and <code>BigMer</code> types for the purposes of writing generalised methods of functions.</p><p>Several aliases are provided for convenience:</p><table><tr><th style="text-align: left">Type alias</th><th style="text-align: left">Type</th></tr><tr><td style="text-align: left"><code>DNAMer{K}</code></td><td style="text-align: left"><code>Mer{DNAAlphabet{2},K}</code></td></tr><tr><td style="text-align: left"><code>RNAMer{K}</code></td><td style="text-align: left"><code>Mer{RNAAlphabet{2},K}</code></td></tr><tr><td style="text-align: left"><code>DNAKmer</code></td><td style="text-align: left"><code>DNAMer{31}</code></td></tr><tr><td style="text-align: left"><code>RNAKmer</code></td><td style="text-align: left"><code>RNAMer{31}</code></td></tr><tr><td style="text-align: left"><code>BigDNAMer{K}</code></td><td style="text-align: left"><code>BigMer{DNAAlphabet{2},K}</code></td></tr><tr><td style="text-align: left"><code>BigRNAMer{K}</code></td><td style="text-align: left"><code>BigMer{RNAAlphabet{2},K}</code></td></tr><tr><td style="text-align: left"><code>BigDNAKmer</code></td><td style="text-align: left"><code>BigMer{DNAAlphabet{2},63}</code></td></tr><tr><td style="text-align: left"><code>BigRNAKmer</code></td><td style="text-align: left"><code>BigMer{RNAAlphabet{2},63}</code></td></tr><tr><td style="text-align: left"><code>DNACodon</code></td><td style="text-align: left"><code>DNAMer{3}</code></td></tr><tr><td style="text-align: left"><code>RNACodon</code></td><td style="text-align: left"><code>RNAMer{3}</code></td></tr></table><h3 id="Skipmers-1"><a class="docs-heading-anchor" href="#Skipmers-1">Skipmers</a><a class="docs-heading-anchor-permalink" href="#Skipmers-1" title="Permalink"></a></h3><p>For some analyses, the contiguous nature of kmers imposes limitations. A single base difference, due to real biological variation or a sequencing error, affects all k-mers crossing that position thus impeding direct analyses by identity. Also, given the strong interdependence of local sequence, contiguous sections capture less information about genome structure, and so they are more affected by sequence repetition. </p><p>Skipmers are a generalisation of the concept of a kmer. They are created using a cyclic pattern of used-and-skipped positions which achieves increased entropy and tolerance to nucleotide substitution differences by following some simple rules.</p><p>Skipmers preserve many of the elegant properties of kmers such as reverse complementability and existence of a canonical representation. Also, using cycles of three greatly increases the power of direct intersection between the genomes of different organisms by grouping together the more conserved  nucleotides of protein-coding regions.</p><p>BioSequences currently does not provide a separate type for skipmers, they are represented using <code>Mer</code> and <code>BigMer</code> as their representation as a short immutable sequence encoded in an unsigned integer is the same. The distinction lies in how they are generated.</p><h4 id="Skipmer-generation-1"><a class="docs-heading-anchor" href="#Skipmer-generation-1">Skipmer generation</a><a class="docs-heading-anchor-permalink" href="#Skipmer-generation-1" title="Permalink"></a></h4><p>A skipmer is a simple cyclic q-gram that includes <em>m</em> out of every <em>n</em> bases until a total of <em>k</em> bases is reached. </p><p>This is illustrated in the figure below (from <a href="https://www.biorxiv.org/content/biorxiv/early/2017/08/23/179960.full.pdf">this paper</a>.):</p><p><img src="../skipmers.png" alt="skipmer-fig"/></p><p>To maintain cyclic properties and the existence of the reverse-complement as a skipmer defined by the same function, <em>k</em> should be a multiple of <em>m</em>.</p><p>This also enables the existence of a canonical representation for each skipmer, defined as the lexicographically smaller of the forward and reverse-complement  representations.</p><p>Defining <em>m</em>, <em>n</em> and <em>k</em> fixes a value for <em>S</em>, the total span of the skipmer, given by: </p><div>\[S = n * (\frac{k}{m} - 1) + m\]</div><p>To see how to iterate over skipmers cf. kmers, see the Iteration section of the manual.</p><h2 id="Reference-sequences-1"><a class="docs-heading-anchor" href="#Reference-sequences-1">Reference sequences</a><a class="docs-heading-anchor-permalink" href="#Reference-sequences-1" title="Permalink"></a></h2><p><code>LongDNASeq</code> (alias of <code>BioSequence{DNAAlphabet{4}}</code>) is a flexible data structure but always consumes 4 bits per base, which will waste a large part of the memory space when storing reference genome sequences.  In such a case, <code>ReferenceSequence</code> is helpful because it compresses positions of &#39;N&#39; symbols so that long DNA sequences are stored with almost 2 bits per base. An important limitation is that the <code>ReferenceSequence</code> type is immutable due to the compression. Other sequence-like operations are supported:</p><pre><code class="language-jlcon">julia&gt; seq = ReferenceSequence(dna&quot;NNCGTATTTTCN&quot;)
12nt Reference Sequence:
NNCGTATTTTCN

julia&gt; seq[1]
DNA_N

julia&gt; seq[5]
DNA_T

julia&gt; seq[2:6]
5nt Reference Sequence:
NCGTA

julia&gt; ReferenceSequence(dna&quot;ATGM&quot;)  # DNA_M is not accepted
ERROR: ArgumentError: invalid symbol M ∉ {A,C,G,T,N} at 4
 in convert at /Users/kenta/.julia/v0.4/Bio/src/seq/refseq.jl:58
 in call at essentials.jl:56
</code></pre><h1 id="Alphabet-types-1"><a class="docs-heading-anchor" href="#Alphabet-types-1">Alphabet types</a><a class="docs-heading-anchor-permalink" href="#Alphabet-types-1" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="BioSequences.Alphabet" href="#BioSequences.Alphabet"><code>BioSequences.Alphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><p><strong>Alphabets of biological symbols.</strong></p><p><code>Alphabet</code> is perhaps the most important type trait for biological sequences in BioSequences.jl.</p><p>An <code>Alphabet</code> represents a domain of biological symbols.</p><p>For example, <code>DNAAlphabet{2}</code> has a domain of unambiguous nucleotides (i.e. A, C, G, and T).</p><p>Alphabet types restrict and define the set of biological symbols, that can be encoded in a given biological sequence type. They ALSO define <em>HOW</em> that encoding is done.</p><p>An <code>Alphabet</code> type defines the encoding of biological symbols with a pair of associated <code>encoder</code> and <code>decoder</code> methods. These paired methods map between biological symbol values and a binary representation of the symbol.</p><p>Any type A &lt;: Alphabet, is expected to implement the <code>Base.eltype</code> method for itself. It is also expected to implement the <code>BitsPerSymbol</code> method.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/93d9462a73a4c156b7069705d0441828dcd2663f/src/alphabet.jl#L10-L32">source</a></section></article><p>Alphabets control how biological symbols are encoded and decoded. They also confer many of the automatic traits and methods that any subtype of <code>T&lt;:BioSequence{A&lt;:Alphabet}</code> will get.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../symbols/">« Biological Symbols</a><a class="docs-footer-nextpage" href="../construction/">Constructing sequences »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 24 June 2020 22:47">Wednesday 24 June 2020</span>. Using Julia version 1.4.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
