<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>BioSequences Types · BioSequences.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="BioSequences.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">BioSequences.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../symbols/">Biological Symbols</a></li><li class="is-active"><a class="tocitem" href>BioSequences Types</a><ul class="internal"><li><a class="tocitem" href="#The-abstract-BioSequence-1"><span>The abstract BioSequence</span></a></li><li><a class="tocitem" href="#The-abstract-Alphabet-1"><span>The abstract Alphabet</span></a></li><li class="toplevel"><a class="tocitem" href="#Concrete-types-1"><span>Concrete types</span></a></li><li><a class="tocitem" href="#Implemented-alphabets-1"><span>Implemented alphabets</span></a></li><li><a class="tocitem" href="#Long-Sequences-1"><span>Long Sequences</span></a></li><li><a class="tocitem" href="#Sequence-views-1"><span>Sequence views</span></a></li></ul></li><li><a class="tocitem" href="../construction/">Constructing sequences</a></li><li><a class="tocitem" href="../transforms/">Indexing &amp; modifying sequences</a></li><li><a class="tocitem" href="../predicates/">Predicates</a></li><li><a class="tocitem" href="../random/">Random sequences</a></li><li><a class="tocitem" href="../sequence_search/">Pattern matching and searching</a></li><li><a class="tocitem" href="../iteration/">Iteration</a></li><li><a class="tocitem" href="../counting/">Counting</a></li><li><a class="tocitem" href="../io/">I/O</a></li><li><a class="tocitem" href="../interfaces/">Interfaces</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>BioSequences Types</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>BioSequences Types</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/BioSequences.jl/blob/master/docs/src/types.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Abstract-Types-1"><a class="docs-heading-anchor" href="#Abstract-Types-1">Abstract Types</a><a class="docs-heading-anchor-permalink" href="#Abstract-Types-1" title="Permalink"></a></h1><p>BioSequences exports an abstract <code>BioSequence</code> type, and several concrete sequence types which inherit from it.</p><h2 id="The-abstract-BioSequence-1"><a class="docs-heading-anchor" href="#The-abstract-BioSequence-1">The abstract BioSequence</a><a class="docs-heading-anchor-permalink" href="#The-abstract-BioSequence-1" title="Permalink"></a></h2><p>BioSequences provides an abstract type called a <code>BioSequence{A&lt;:Alphabet}</code>. This abstract type, and the methods and traits is supports, allows for many algorithms in BioSequences to be written as generically as possible, thus reducing the amount of code to read and understand, whilst maintaining high performance when such code is compiled for a concrete BioSequence subtype. Additionally, it allows new types to be implemented that are fully compatible with the rest of BioSequences, providing that key methods or traits are defined).</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.BioSequence" href="#BioSequences.BioSequence"><code>BioSequences.BioSequence</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">BioSequence{A &lt;: Alphabet}</code></pre><p><code>BioSequence</code> is the main abstract type of <code>BioSequences</code>. Its subtypes are characterized by:</p><ul><li>Being a linear container type with random access and indices <code>Base.OneTo(length(x))</code>.</li><li>Containing zero or more internal data elements of type <code>encoded_data_eltype(typeof(x))</code>.</li><li>Being associated with an <code>Alphabet</code>, <code>A</code> by being a subtype of <code>BioSequence{A}</code>.</li></ul><p>A <code>BioSequence{A}</code> is indexed by an integer. The biosequence subtype, the index and the alphabet <code>A</code> determine how to extract the internal encoded data. The alphabet decides how to decode the data to the element type of the biosequence. Hence, the element type and container type of a <code>BioSequence</code> are separated.</p><p>Subtypes <code>T</code> of <code>BioSequence</code> must implement the following, with <code>E</code> begin an encoded data type:</p><ul><li><code>Base.length(::T)::Int</code></li><li><code>encoded_data_eltype(::Type{T})::Type{E}</code></li><li><code>extract_encoded_element(::T, ::Integer)::E</code></li><li><code>copy(::T)</code></li><li>T must be able to be constructed from any iterable with <code>length</code> defined and</li></ul><p>with a known, compatible element type.</p><p>Furthermore, mutable sequences should implement</p><ul><li><code>encoded_setindex!(::T, ::E, ::Integer)</code></li><li><code>T(undef, ::Int)</code></li><li>resize!(::T, ::Int)</li></ul><p>For compatibility with existing <code>Alphabet</code>s, the encoded data eltype must be <code>UInt</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/biosequence/biosequence.jl#L7-L37">source</a></section></article><p>Some aliases for <code>BioSequence</code> are also provided for your convenience:</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.NucleotideSeq" href="#BioSequences.NucleotideSeq"><code>BioSequences.NucleotideSeq</code></a> — <span class="docstring-category">Type</span></header><section><div><p>An alias for <code>BioSequence{&lt;:NucleicAcidAlphabet}</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/biosequence/biosequence.jl#L128-L130">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.AminoAcidSeq" href="#BioSequences.AminoAcidSeq"><code>BioSequences.AminoAcidSeq</code></a> — <span class="docstring-category">Type</span></header><section><div><p>An alias for <code>BioSequence{AminoAcidAlphabet}</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/biosequence/biosequence.jl#L133-L135">source</a></section></article><p>Let&#39;s have a closer look at some of those methods that a subtype of <code>BioSequence</code> must implement. Check out julia base library docs for <code>length</code>, <code>copy</code> and <code>resize!</code>.</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.encoded_data_eltype" href="#BioSequences.encoded_data_eltype"><code>BioSequences.encoded_data_eltype</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">encoded_data_eltype(::Type{&lt;:BioSequence})</code></pre><p>Returns the element type of the encoded data of the <code>BioSequence</code>. This is the return type of <code>extract_encoded_element</code>, i.e. the data type that stores the biological symbols in the biosequence.</p><p>See also: <a href="#BioSequences.BioSequence"><code>BioSequence</code></a> </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/biosequence/biosequence.jl#L94-L102">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.extract_encoded_element" href="#BioSequences.extract_encoded_element"><code>BioSequences.extract_encoded_element</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">extract_encoded_element(::BioSequence{A}, i::Integer)</code></pre><p>Returns the encoded element at position <code>i</code>. This data can be decoded using <code>decode(A(), data)</code> to yield the element type of the biosequence.</p><p>See also: <a href="#BioSequences.BioSequence"><code>BioSequence</code></a> </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/biosequence/biosequence.jl#L105-L113">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.encoded_setindex!" href="#BioSequences.encoded_setindex!"><code>BioSequences.encoded_setindex!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">encoded_setindex!(seq::BioSequence, x::E, i::Integer)</code></pre><p>Given encoded data <code>x</code> of type <code>encoded_data_eltype(typeof(seq))</code>, sets the internal sequence data at the given index.</p><p>See also: <a href="#BioSequences.BioSequence"><code>BioSequence</code></a> </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/biosequence/biosequence.jl#L117-L124">source</a></section></article><p>A correctly defined subtype of <code>BioSequence</code> that satisfies the interface, will find the vast majority of methods described in the rest of this manual should work out of the box for that type. But they can always be overloaded if needed. Indeed the <code>LongSequence</code> type overloads Indeed some of the generic <code>BioSequence</code> methods, are overloaded for <code>LongSequence</code>, for example for transformation and counting operations where efficiency gains can be made due to the specific internal representation of a specific type.</p><h2 id="The-abstract-Alphabet-1"><a class="docs-heading-anchor" href="#The-abstract-Alphabet-1">The abstract Alphabet</a><a class="docs-heading-anchor-permalink" href="#The-abstract-Alphabet-1" title="Permalink"></a></h2><p>Alphabets control how biological symbols are encoded and decoded. They also confer many of the automatic traits and methods that any subtype of <code>T&lt;:BioSequence{A&lt;:Alphabet}</code> will get.</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.Alphabet" href="#BioSequences.Alphabet"><code>BioSequences.Alphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><p><strong>Alphabets of biological symbols.</strong></p><p><code>Alphabet</code> is the most important type trait for <code>BioSequence</code>. An <code>Alphabet</code> represents a set of biological symbols encoded by a sequence, e.g. A, C, G and T for a DNA Alphabet that requires only 2-bits to represent each symbol.</p><ul><li>Subtypes of Alphabet are singleton structs that may or may not be parameterized.</li><li>Alphabets span over a <em>finite</em> set of biological symbols.</li><li>The alphabet controls the encoding from some internal &quot;encoded data&quot; to a BioSymbol </li></ul><p>of the alphabet&#39;s element type, as well as the decoding, the inverse process.</p><ul><li>An <code>Alphabet</code>&#39;s <code>encode</code> and <code>decode</code> methods must not produce invalid data. </li></ul><p>Every subtype <code>A</code> of <code>Alphabet</code> must implement:</p><ul><li><code>Base.eltype(::Type{A})::Type{S}</code> for some eltype <code>S</code>, which must be a BioSymbol.</li><li><code>symbols(::A)::Tuple{Vararg{S}}</code>. This gives tuples of all symbols in the set of <code>A</code>.</li><li><code>encode(::A, ::S)::E</code> encodes a symbol to an internal data eltype <code>E</code>.</li><li><code>decode(::A, ::E)::S</code> decodes an internal data eltype <code>E</code> to a symbol <code>S</code>.</li><li>Except for <code>eltype</code> which must follow Base conventions, all functions operating</li></ul><p>on <code>Alphabet</code> should operate on instances of the alphabet, not the type.</p><p>If you want interoperation with existing subtypes of <code>BioSequence</code>, the encoded representation <code>E</code> must be of type <code>UInt</code>, and you must also implement:</p><ul><li><code>BitsPerSymbol(::A)::BitsPerSymbol{N}</code>, where the <code>N</code> must be zero</li></ul><p>or a power of two in [1, 2, 4, 8, 16, 32, [64 for 64-bit systems]].</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/fa8828b69e9c72dd4df03994a73527576bd0e13d/src/alphabet.jl#L10-L35">source</a></section></article><h1 id="Concrete-types-1"><a class="docs-heading-anchor" href="#Concrete-types-1">Concrete types</a><a class="docs-heading-anchor-permalink" href="#Concrete-types-1" title="Permalink"></a></h1><h2 id="Implemented-alphabets-1"><a class="docs-heading-anchor" href="#Implemented-alphabets-1">Implemented alphabets</a><a class="docs-heading-anchor-permalink" href="#Implemented-alphabets-1" title="Permalink"></a></h2><h2 id="Long-Sequences-1"><a class="docs-heading-anchor" href="#Long-Sequences-1">Long Sequences</a><a class="docs-heading-anchor-permalink" href="#Long-Sequences-1" title="Permalink"></a></h2><p>Many genomics scripts and tools benefit from an efficient general purpose sequence type that allows you to create and edit sequences. In BioSequences, the <code>LongSequence</code> type fills this requirement.</p><p><code>LongSequence{A&lt;:Alphabet} &lt;: BioSequence{A}</code> is parameterized by a concrete <code>Alphabet</code> type <code>A</code> that defines the domain (or set) of biological symbols permitted. For example, <code>AminoAcidAlphabet</code> is associated with <code>AminoAcid</code> and hence an object of the <code>LongSequence{AminoAcidAlphabet}</code> type represents a sequence of amino acids.  Symbols from multiple alphabets can&#39;t be intermixed in one sequence type.</p><p>The following table summarizes common LongSequence types that have been given aliases for convenience.</p><table><tr><th style="text-align: left">Type</th><th style="text-align: left">Symbol type</th><th style="text-align: left">Type alias</th></tr><tr><td style="text-align: left"><code>LongSequence{DNAAlphabet{4}}</code></td><td style="text-align: left"><code>DNA</code></td><td style="text-align: left"><code>LongDNASeq</code></td></tr><tr><td style="text-align: left"><code>LongSequence{RNAAlphabet{4}}</code></td><td style="text-align: left"><code>RNA</code></td><td style="text-align: left"><code>LongRNASeq</code></td></tr><tr><td style="text-align: left"><code>LongSequence{AminoAcidAlphabet}</code></td><td style="text-align: left"><code>AminoAcid</code></td><td style="text-align: left"><code>LongAminoAcidSeq</code></td></tr></table><p>The <code>LongDNASeq</code> and <code>LongRNASeq</code> aliases use a DNAAlphabet{4}, which means the sequence may store ambiguous nucleotides. If you are sure that nucleotide sequences store unambiguous nucleotides only, you can reduce the memory required by sequences by using a slightly different parameter: <code>DNAAlphabet{2}</code> is an alphabet that uses two bits per base and limits to only unambiguous nucleotide symbols (ACGT in DNA and ACGU in RNA). Replacing <code>LongSequence{DNAAlphabet{4}}</code> in your code with <code>LongSequence{DNAAlphabet{2}}</code> is all you need to do in order to benefit. Some computations that use bitwise operations will also be dramatically faster.</p><h2 id="Sequence-views-1"><a class="docs-heading-anchor" href="#Sequence-views-1">Sequence views</a><a class="docs-heading-anchor-permalink" href="#Sequence-views-1" title="Permalink"></a></h2><p>Similar to how Base Julia offers views of array objects, BioSequences offers view of <code>LongSequence</code>s - the <code>LongSubSeq{A&lt;:Alphabet}</code>.</p><p>Conceptually, a <code>LongSubSeq{A}</code> is similar to a <code>LongSequence{A}</code>, but instead of storing their own data, they refer to the data of a <code>LongSequence</code>. Modiying the <code>LongSequence</code> will be reflected in the view, and vice versa. If the underlying <code>LongSequence</code> is truncated, the behaviour of a view is undefined. For the same reason, some operations are not supported for views, such as resizing.</p><p>The purpose of <code>LongSubSeq</code> is that, since they only contain a pointer to the underlying array, an offset and a length, they are much lighter than <code>LongSequences</code>, and will be stack allocated on Julia 1.5 and newer. Thus, the user may construct millions of views without major performance implications.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../symbols/">« Biological Symbols</a><a class="docs-footer-nextpage" href="../construction/">Constructing sequences »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 18 June 2021 18:49">Friday 18 June 2021</span>. Using Julia version 1.6.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
