<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Nucleic acid k-mers · BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../index.html"><img class="logo" src="../assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../index.html">Home</a></li><li><a class="toctext" href="../symbols.html">Biological Symbols</a></li><li><span class="toctext">Sequence types</span><ul><li><a class="toctext" href="sequences.html">Overview</a></li><li><a class="toctext" href="bioseq.html">BioSequence</a></li><li><a class="toctext" href="refseq.html">Reference Sequences</a></li><li class="current"><a class="toctext" href="kmer.html">Nucleic acid k-mers</a><ul class="internal"><li><a class="toctext" href="#Nucleic-acid-k-mers-1">Nucleic acid k-mers</a></li></ul></li></ul></li><li><span class="toctext">IO</span><ul><li><a class="toctext" href="../io/fasta.html">FASTA formatted files</a></li><li><a class="toctext" href="../io/fastq.html">FASTQ formatted files</a></li><li><a class="toctext" href="../io/twobit.html">2bit formatted files</a></li></ul></li><li><a class="toctext" href="../search.html">Searching</a></li><li><a class="toctext" href="../composition.html">Sequence Composition</a></li><li><a class="toctext" href="../demultiplexer.html">Demultiplexing</a></li><li><a class="toctext" href="../contributing.html">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li>Sequence types</li><li><a href="kmer.html">Nucleic acid k-mers</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/BioSequences.jl/blob/master/docs/src/sequences/kmer.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Nucleic acid k-mers</span><a class="fa fa-bars" href="#"></a></div></header><h2><a class="nav-anchor" id="Nucleic-acid-k-mers-1" href="#Nucleic-acid-k-mers-1">Nucleic acid k-mers</a></h2><p>A common strategy to simplify the analysis of sequence data is to operate or short k-mers, for size fixed size <code>k</code>. These can be packed into machine integers allowing extremely efficient code. The <code>BioSequences</code> module has built in support for representing short sequences in 64-bit integers. Besides being fixed length, <code>Kmer</code> types, unlike other sequence types cannot contain ambiguous symbols like &#39;N&#39;.</p><p>The <code>Kmer{T,k}</code> type parameterized on symbol type (<code>T</code>, either <code>DNA</code>, or <code>RNA</code>) and size <code>k</code>. For ease of writing code, two type aliases for each nucleotide type are defined and named as <code>DNAKmer{k}</code> and <code>RNAKmer{k}</code>:</p><pre><code class="language-julia-repl">julia&gt; DNAKmer(&quot;ACGT&quot;)  # create a DNA 4-mer from a string
DNA 4-mer:
ACGT

julia&gt; RNAKmer(&quot;ACGU&quot;)  # create an RNA 4-mer from a string
RNA 4-mer:
ACGU

julia&gt; kmer&quot;ACGT&quot; # DNA k-mers may also be written as literals
DNA 4-mer:
ACGT

julia&gt; typeof(DNAKmer(&quot;ACGT&quot;))
BioSequences.Kmer{BioSymbols.DNA,4}</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.each" href="#BioSequences.each"><code>BioSequences.each</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">each(::Type{Kmer{T,k}}, seq::Sequence[, step=1])</code></pre><p>Initialize an iterator over all k-mers in a sequence <code>seq</code> skipping ambiguous nucleotides without changing the reading frame.</p><p><strong>Arguments</strong></p><ul><li><p><code>Kmer{T,k}</code>: k-mer type to enumerate.</p></li><li><p><code>seq</code>: a nucleotide sequence.</p></li><li><p><code>step=1</code>: the number of positions between iterated k-mers</p></li></ul><p><strong>Examples</strong></p><pre><code class="language-none"># iterate over DNA codons
for (pos, codon) in each(DNAKmer{3}, dna&quot;ATCCTANAGNTACT&quot;, 3)
    @show pos, codon
end</code></pre></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/bec2ae2001c5fc943214d0ceafd2519d11dcd865/src/eachkmer.jl#L16-L34">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.canonical" href="#BioSequences.canonical"><code>BioSequences.canonical</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">canonical(kmer::Kmer)</code></pre><p>Return the canonical k-mer of <code>x</code>.</p><p>A canonical k-mer is the numerical lesser of a k-mer and its reverse complement. This is useful in hashing/counting k-mers in data that is not strand specific, and thus observing k-mer is equivalent to observing its reverse complement.</p></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/bec2ae2001c5fc943214d0ceafd2519d11dcd865/src/kmer.jl#L194-L202">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.neighbors" href="#BioSequences.neighbors"><code>BioSequences.neighbors</code></a> — <span class="docstring-category">Function</span>.</div><div><pre><code class="language-none">neighbors(kmer::Kmer)</code></pre><p>Return an iterator through k-mers neighboring <code>kmer</code> on a de Bruijn graph.</p></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioSequences.jl/blob/bec2ae2001c5fc943214d0ceafd2519d11dcd865/src/kmer.jl#L225-L229">source</a></section><footer><hr/><a class="previous" href="refseq.html"><span class="direction">Previous</span><span class="title">Reference Sequences</span></a><a class="next" href="../io/fasta.html"><span class="direction">Next</span><span class="title">FASTA formatted files</span></a></footer></article></body></html>
