<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Sequence composition · BioSequences.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="BioSequences.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">BioSequences.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../symbols/">Biological Symbols</a></li><li><a class="tocitem" href="../types/">BioSequences Types</a></li><li><a class="tocitem" href="../construction/">Constructing sequences</a></li><li><a class="tocitem" href="../transforms/">Indexing &amp; modifying sequences</a></li><li><a class="tocitem" href="../predicates/">Predicates</a></li><li><a class="tocitem" href="../random/">Random sequences</a></li><li><a class="tocitem" href="../sequence_search/">Pattern matching and searching</a></li><li><a class="tocitem" href="../demultiplexer/">Sequence Demultipllexing</a></li><li class="is-active"><a class="tocitem" href>Sequence composition</a></li><li><a class="tocitem" href="../iteration/">Iteration</a></li><li><a class="tocitem" href="../counting/">Counting</a></li><li><a class="tocitem" href="../io/">I/O</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Sequence composition</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Sequence composition</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/BioSequences.jl/blob/master/docs/src/composition.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Sequence-composition-1"><a class="docs-heading-anchor" href="#Sequence-composition-1">Sequence composition</a><a class="docs-heading-anchor-permalink" href="#Sequence-composition-1" title="Permalink"></a></h1><p>There are many instances in analyzing sequence data where you will want to know about the composition of your sequences.</p><p>For example, for a given sequence, you may want to count how many of each possible Kmer, is present in the sequence. This would be important if - for instance - you wanted to analyze the Kmer spectra of your data. Alternatively you might have a collection of sequences, and may want to count how many of each unique sequence you have in your collection. This would be important if - for instance - your collection of sequences were from a population sample, and you wanted to compute the allele or genotype frequencies for the population.</p><p>Whatever the application, BioSequences provides a method called <code>composition</code>, and a parametric struct called <code>Composition</code> to both compute, and handle the results of such sequence composition calculations.</p><article class="docstring"><header><a class="docstring-binding" id="BioSequences.Composition" href="#BioSequences.Composition"><code>BioSequences.Composition</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Sequence composition.</p><p>This is a subtype of <code>Associative{T,Int}</code>, and the <code>getindex</code> method returns the number of occurrences of a symbol or a k-mer.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="BioSequences.composition" href="#BioSequences.composition"><code>BioSequences.composition</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">composition(seq | kmer_iter)</code></pre><p>Calculate composition of biological symbols in <code>seq</code> or k-mers in <code>kmer_iter</code>.</p></div></section><section><div><pre><code class="language-none">composition(iter)</code></pre><p>A generalised composition algorithm, which computes the number of unique items produced by an iterable.</p><p><strong>Example</strong></p><pre><code class="language-jlcon">
# Example, counting unique sequences.

julia&gt; a = dna&quot;AAAAAAAATTTTTT&quot;
14nt DNA Sequence:
AAAAAAAATTTTTT

julia&gt; b = dna&quot;AAAAAAAATTTTTT&quot;
14nt DNA Sequence:
AAAAAAAATTTTTT

julia&gt; c = a[5:10]
6nt DNA Sequence:
AAAATT

julia&gt; composition([a, b, c])
Vector{BioSequences.BioSequence{BioSequences.DNAAlphabet{4}}} Composition:
  AAAATT         =&gt; 1
  AAAAAAAATTTTTT =&gt; 2</code></pre></div></section></article><p>For example to get the nucleotide composition of a sequence:</p><pre><code class="language-julia-repl">julia&gt; comp = composition(dna&quot;ACGAG&quot;);

julia&gt; comp[DNA_A]
2

julia&gt; comp[DNA_T]
0
</code></pre><p>Composition structs behave like an associative collection, such as a <code>Dict</code>. But there are a few differences:</p><ol><li>The <code>getindex</code> method for Composition structs is overloaded to return a default value of 0, if a key is used that is not present in the Composition.</li><li>The <code>merge!</code> method for two Composition structs adds counts together, unlike the <code>merge!</code> method for other associative containers, which would overwrite the counts.</li></ol><p><code>merge!</code> is used to accumulate composition statistics of multiple sequences:</p><pre><code class="language-julia-repl">julia&gt; # initiaize an empty composition counter
       comp = composition(dna&quot;&quot;);
ERROR: LoadError: UndefVarError: @dna_str not defined
in expression starting at none:2

julia&gt; # iterate over sequences and accumulate composition statistics into `comp`
       for seq in seqs
           merge!(comp, composition(seq))
       end
ERROR: UndefVarError: seqs not defined

julia&gt; # or functional programming style in one line
       foldl((x, y) -&gt; merge(x, composition(y)), composition(dna&quot;&quot;), seqs)
ERROR: LoadError: UndefVarError: @dna_str not defined
in expression starting at none:3</code></pre><p><code>composition</code> is also applicable to a <em>k</em>-mer iterator:</p><pre><code class="language-julia-repl">julia&gt; comp = composition(each(DNAMer{4}, dna&quot;ACGT&quot;^100));

julia&gt; comp[DNAMer(&quot;ACGT&quot;)]
100

julia&gt; comp[DNAMer(&quot;CGTA&quot;)]
99
</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../demultiplexer/">« Sequence Demultipllexing</a><a class="docs-footer-nextpage" href="../iteration/">Iteration »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 20 January 2020 11:22">Monday 20 January 2020</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
